src/main/scala/mesosphere/marathon/core/launchqueue/impl/ReviveOffersStreamLogic.scala

```diff
@@ -34,115 +38,221 @@ object ReviveOffersStreamLogic extends StrictLogging {
+    Flow[Map[Role, VersionedRoleState]]
+      .sliding(2)
+      .mapConcat({
+        case Seq(lastState, newState) =>
+          logic.directivesForDiff(lastState, newState)
+        case _ =>
+          logger.debug(s"Stream is terminating")
+          Nil
+      })
+  }
 
   /**
-    * This flow emits a [[Suppress]] or a [[Revive]] based on the last two states emitted by [[reviveStateFromInstancesAndDelays]]
+    * Immutable directive generator which compares two offers wanted state and issues the appropriate unsuppress or
+    * re-revive directives.
     *
-    * @return The actual flow.
+    * There are two implementations for the logic, one with suppression, and the other with suppression disabled.
     */
-  val suppressOrReviveFromDiff: Flow[ReviveOffersState, Op, NotUsed] = Flow[ReviveOffersState]
-    .sliding(2)
-    .mapConcat {
-      case Seq(previous, current) =>
-
-        val diffScheduled = current.scheduledInstancesWithoutBackoff -- previous.scheduledInstancesWithoutBackoff
-        def diffTerminal = current.terminalReservations -- previous.terminalReservations
-
-        if (diffScheduled.nonEmpty) {
-          logger.info(s"Revive: There are newly scheduled instances to launch: $diffScheduled")
-          // There's a very small chance that we decline an offer in response to a revive for an instance not yet registered
-          // with the TaskLauncherActor. To deal with the rare case this happens, we just repeat the revive call a few times
-          List(Revive, Revive, Revive)
-        } else if (diffTerminal.nonEmpty) {
-          logger.info(s"Revive: There are decommissioned resident instances that need their reservations cleaned up: $diffTerminal")
-          List(Revive)
-        } else if (shouldTryToReleaseForceExpungedResidentInstances(current, previous)) {
-          logger.info(s"Revive: Triggering reservation reconciliation.")
-          List(Revive)
-        } else if (current.isEmpty) {
-          logger.info(s"Suppress: There are no pending instances right now.")
-          List(Suppress)
-        } else {
-          Nil
+  private[impl] trait ReviveDirectiveFlowLogic {
+    def lastOffersWantedVersion(lastState: Map[Role, VersionedRoleState], role: Role): Option[Long] =
+      lastState.get(role).collect { case VersionedRoleState(version, OffersWanted) => version }
+
+    def directivesForDiff(lastState: Map[Role, VersionedRoleState], newState: Map[Role, VersionedRoleState]): List[RoleDirective]
+  }
+
+  private[impl] class ReviveDirectiveFlowLogicWithoutSuppression extends ReviveDirectiveFlowLogic {
+
+    def directivesForDiff(lastState: Map[Role, VersionedRoleState], newState: Map[Role, VersionedRoleState]): List[RoleDirective] = {
+      val rolesChanged = lastState.keySet != newState.keySet
+      val directives = List.newBuilder[RoleDirective]
+
+      if (rolesChanged) {
+        val newRoleState = newState.keysIterator.map { role =>
```

Comment:
I'm sorry. I don't get this. Why are we mapping to `OfferWanted` for everything? Is it because we never suppress?

Reply:
Yes